Created DELETE /groups/{groupId} and PUT /groups/{groupId}/leave mapp…

OrestisBollano opened the pull request:
The name of the branch is a bit misleading as well tbh since I was waiting for the review for the original PR started working on other issues. But you can find the number of the issue attached on the commit 🍰

wlany commented on src/main/java/ch/uzh/ifi/hase/soprafs23/service/UserService.java:
```diff
@@ -208,4 +211,20 @@ public void addUserIngredients(Long userId, List<IngredientPutDTO> ingredientsPu
+        Long groupId = user.getGroupId();
+        if (groupId != null) {
+            Group group = groupRepository.findById(groupId).orElseThrow(() ->
+                    new ResponseStatusException(HttpStatus.NOT_FOUND, String.format("Group with id %d does not exist", groupId)));
```
I think it'd be better if you let UserService have access to GroupService instead of GroupRepository.
Then Group group = getGroupById(groupId); would do this 404 for you (no need for the orElseThrow()), and also we would have the Repositories only being user by their corresponding Services -> a bit less coupling, and the Services have all the functionality needed

wlany commented on src/main/java/ch/uzh/ifi/hase/soprafs23/controller/GroupController.java:
```diff
@@ -223,5 +223,50 @@ public List<UserGetDTO> getGroupMembersById(@PathVariable Long groupId, HttpServ
+        if (!memberIds.contains(guestId)) {
+            throw new ResponseStatusException(HttpStatus.CONFLICT, "Guest was not in the group"); // 409 - conflict
+        }
+
```
PROBLEM! check if the tokenId corresponds to the hostId of the group! atm the host can leave the group and it's very weird because in the group they are still the host but in the users they are in no group and all the guests stay in the group ^^' 
I'd say throw a 409 in that case with something like "you are the host. if you want to leave please delete the group" or something like that

akkuebler replied:
like the error suggestion!

wlany reviewed (changes requested):
tested it with Postman as well and works :D
one big bug though: the host can leave the group ^^'

akkuebler commented on src/main/java/ch/uzh/ifi/hase/soprafs23/controller/GroupController.java:
```diff
@@ -223,5 +223,50 @@ public List<UserGetDTO> getGroupMembersById(@PathVariable Long groupId, HttpServ
+
+        // Check if the user is the host of the group
+        if (!tokenId.equals(group.getHostId())) {
+            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Not authorized"); // 401 - not authorized
```
can you dd here sth like: Only the host of the group can delete the group?

wlany reviewed (commented):
It's so much code at this point, I feel I have no overview over the code ^^' I tested with postman, here my thoughts:
- atm a user already in a group can request to join a group -> should receive some warning "you're already in a group" or something
- I don't really understand what the join request status is about... when it's rejected for example I can still go and accept it, if it's either (accepted or rejected) and I post the same request again it's just set back to pending (if it was accepted it's a bit weird because then the status is pending but the user is in the group), shouldn't it alert me that the request is already accepted/rejected? otherwhise I don't see a reason for the status, and think making it analogue to the invitations, such that upon accepting/rejecting the invitation is deleted would be better... 
- the GET request to get all open requests to join a certain group is missing (but idk if this is another task you haven't gotten to :))


maybe we can talk about this this afternoon/evening :)

OrestisBollano commented:
Updated the 3 methods of the POST /groups/{groupId}/request, PUT /groups/{groupId}/request/accept, and PUT /groups/{groupId}/request/reject.

POST /groups/{groupId}/request: Added the FORBIDDEN status check so that the user that is already in a group cannot make a join request for another group.
PUT /groups/{groupId}/request/accept: When a user's join request to a group is accepted all his other join requests are deleted form the database.
PUT /groups/{groupId}/request/reject.: When a user's join request to a group is rejected then only that join request is deleted while the other join requests that he has made are still active and pending.

Also removed the JoinRequestStatus as we discussed. 
P.S. Also with POSTMAN and the database I tried a couple of logic between the invitation and a join request. e.g. when a user has an active join request for a group and an invitation for that group at the same time and they seem to work well, when the invitation is accepted then the join request returns a status code. Feel free to also try some things before you can merge this branch :)

wlany commented on src/main/java/ch/uzh/ifi/hase/soprafs23/service/JoinRequestService.java:
```diff
@@ -0,0 +1,87 @@
+    public JoinRequest createJoinRequest(JoinRequestPostDTO joinRequestPostDTO, Long groupId) {
+        Optional<JoinRequest> existingJoinRequest = joinRequestRepository.findByGuestIdAndGroupId(joinRequestPostDTO.getGuestId(), groupId);
+        if (existingJoinRequest.isPresent()) {
+            throw new ResponseStatusException(HttpStatus.CONFLICT, "A request could not be sent");
```
this message could be more specific, something like "guest X already has an open request to join group Y"

wlany reviewed (commented):
I tested it a lot in postman it works perfect :D 
the only thing I noticed is the one error message that could be a bit more specific, as the user would then know that their request exists

wlany commented on src/main/java/ch/uzh/ifi/hase/soprafs23/service/JoinRequestService.java:
```diff
@@ -0,0 +1,98 @@
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Join request not found"); // 404 - join request not found
+        }
+
+        JoinRequest joinRequest = joinRequestOptional.get();
```
This duplicate code has not been addressed yet

akkuebler reviewed (approved):
nice adaptation of comments - thanks!
